### PR TITLE
gh-94471: Enable pointer authentication on aarch64 builds

### PR DESCRIPTION
**Pull request by jgowdy**
ARM64v8.3 supports [Pointer Authentication](https://www.qualcomm.com/media/documents/files/whitepaper-pointer-authentication-on-armv8-3.pdf) with the PACIASP and AUTIASP instructions which are interpreted as NOP instructions on pre-8.3 architectures.  These instructions sign the stack pointer and validate the stack pointer prior to return to mitigate [return oriented programming](https://en.wikipedia.org/wiki/Return-oriented_programming).

GCC supports these [options](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html) on arm64 / aarch64.  The legacy option was `-msign-return-address=[all | non-leaf | none]` and the modern option is `-mbranch-protection=none|standard|pac-ret[+leaf+b-key]|bti`

I would like to suggest that the arm64 build be modified to include `-mbranch-protection=pac-ret` with the `-march` being set to ARMv8.2 or earlier or not configured, so that GCC will generate [PACIASP](https://developer.arm.com/documentation/100076/0100/A64-Instruction-Set-Reference/A64-General-Instructions/PACIA--PACIZA--PACIA1716--PACIASP--PACIAZ?lang=en) and [AUTIASP](https://developer.arm.com/documentation/100076/0100/A64-Instruction-Set-Reference/A64-General-Instructions/AUTIA--AUTIZA--AUTIA1716--AUTIASP--AUTIAZ?lang=en) instructions.  It is critical that `-march=armv8.3` or higher not be passed or the non-backwards compatible RETAA instruction will be generated.

The benefit of enabling pointer authentication for the stack pointer on ARM64 would be to mitigate [return oriented programming](https://en.wikipedia.org/wiki/Return-oriented_programming) attacks against the CPython runtime.

Presently we (GoDaddy) are pursuing custom compiles of the CPython runtime for the new Graviton3 CPUs that support pointer authentication in AWS.

* Issue: gh-94471

### CODE REVIEW

**Comment by jgowdy** [comment]
I'll need to regenerate configure with the appropriate version of autoconf

**Comment by erlend-aasland** [comment]
> I'll need to regenerate configure with the appropriate version of autoconf

`docker run --rm --pull=always -v"$PWD:/src" [quay.io/tiran/cpython_autoconf:269](http://quay.io/tiran/cpython_autoconf:269)`

**Comment by jgowdy** [comment]
@erlend-aasland Looks green 👍🏻

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: I wonder if we should write out the `for` loop instead. AFAIK, `AS_FOR` is undocumented.

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: Please use `AS_VAR_IF` here.

**Comment by jgowdy** [comment]
I've attempted to simplify the M4sh here to only need AS_CASE and AX_CHECK_COMPILE_FLAG.  Is this suitable?

**Comment by erlend-aasland** [comment]
Thanks. The AC change looks good to me. I have no opinions regarding the arch specifics here. Perhaps @corona10 or @tiran can chime in for that?

**Comment by tiran** [comment]
Are there potential performance issues or backwards compatibility issues with `-mbranch-protection`?

`-msign-return-address` is a legacy option. It looks like GCC 9 and newer have `-mbranch-protection`. GCC 9 or newer are available for most distros, even RHEL 7 with DevTool set.

**Comment by jgowdy** [comment]
> Are there potential performance issues or backwards compatibility issues with `-mbranch-protection`?
> 
> `-msign-return-address` is a legacy option. It looks like GCC 9 and newer have `-mbranch-protection`. GCC 9 or newer are available for most distros, even RHEL 7 with DevTool set.

I should have also stated, there isn't a backwards compatibility issue here since the instructions are implemented in the NOP space for prior architectures.

**Review by erlend-aasland — APPROVED** [review]
LGTM, but I would like @corona10 or another core dev to approve before merging.

**Comment by corona10** [comment]
@erlend-aasland I will take a look at this PR, I may need to understand potential impact with this patch :)

**Comment by corona10** [comment]
@pablogsal cc @erlend-aasland 
Is this patch will affect trampoline implementation?
Can you please check the impact?
https://github.com/python/cpython/blob/1f737edb67e702095feb97118a911afb569f5705/Python/asm_trampoline.S#L17-L25

@jgowdy Would you like to submit the pyperformance benchmark result comparing the result between
non-pointer authentication  and pointer authentication.

**Comment by pablogsal** [comment]
> @pablogsal cc @erlend-aasland 
> 
> Is this patch will affect trampoline implementation?
> 
It has potential to, specifically it may affect unwinders (including perf), debuggers and state inspection tools. We should check if the main unwinders know how to handle these pointers correctly.

**Comment by pablogsal** [comment]
I'm also checking with the buildbot fleet.

**Comment by pablogsal** [comment]
From https://lore.kernel.org/linux-arm-kernel/Y1q914IVy6XgE1xq@hirez.programming.kicks-ass.net/t/ :

> Perf report cannot produce callgraphs using dwarf on arm64 where pointer
authentication is enabled.  This is because libunwind and libdw cannot
unmangle instruction pointers that have a pointer authentication code
(PAC) embedded in them.
> libunwind and libdw need to be given an instruction mask which they can
use to arrive at the correct return address that does not contain the
PAC.


I think, if confirmed, this is a significant enough drawback that makes this a big blocker for merging this.

**Comment by erlend-aasland** [comment]
Thanks, Dong-hee and Pablo! I'll mark this as do-not-merge, just to be on the safe side.

**Comment by pablogsal** [comment]
This makes `test_gdb` fail on aarch64 RHEL buildbots, so this more or less confirms that this messes up with debuggers as I suspected.

**Comment by erlend-aasland** [comment]
Given Dong-hee's and Pablo's valuable input on this, I believe we can conclude with that this feature is not something we can easily implement. Suggesting to close this PR and the linked issue.

**Comment by pablogsal** [comment]
Agreed, also this is something that redistributors can easily do by setting CFLAGS or similar on they custom builds if they are ok sacrificing debugging / profilings capabilities.

**Pull request closed without merging**
